clang/test/SemaSYCL/sycl-external-attr.cpp

```diff
@@ -0,0 +1,70 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+// RUN: %clang_cc1 -fsycl-is-device -std=c++20 -fsyntax-only -verify -DCPP20 %s
+// Semantic tests for sycl_external attribute
+
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+static void func1() {}
+
+namespace {
+  [[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func2() {}
+
+  struct UnnX {};
+}
+
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func4(UnnX) {}
+
+// The first declaration of a SYCL external function is required to have this attribute.
+int foo(); // expected-note {{previous declaration is here}}
+
+[[clang::sycl_external]] int foo(); // expected-error {{'clang::sycl_external' attribute does not appear on the first declaration}}
+
+// Subsequent declrations of a SYCL external function may optionally specify this attribute.
+[[clang::sycl_external]] int boo();
+
+[[clang::sycl_external]] int boo(); // OK
+
+int boo(); // OK
+
+class C {
+  [[clang::sycl_external]] void member();
+};
+
+[[clang::sycl_external]] int main() // expected-error {{'sycl_external' cannot be applied to the 'main' function}}
+{
+  return 0;
+}
+
+class D {
+  [[clang::sycl_external]] void del() = delete; // expected-error {{'sycl_external' cannot be applied to an explicitly deleted function}}
+};
+
+struct NonCopyable {
+  ~NonCopyable() = delete;
+  [[clang::sycl_external]] NonCopyable(const NonCopyable&) = default;
+};
+
+class A {
+  [[clang::sycl_external]]
+  A() {}
+
+  [[clang::sycl_external]] void func3() {}
+};
+
+class B {
+public:
+  [[clang::sycl_external]] virtual void foo() {}
+
+  [[clang::sycl_external]] virtual void bar() = 0;
+};
+
+[[clang::sycl_external]] int *func0() { return nullptr; }
+
+[[clang::sycl_external]] void func2(int *) {}
```

Comment:
I assume these are intended to exercise the generic address space limitations. Let's add a comment to make that clear. I don't know if any of the target devices we expect to support lack support for the generic address space, so we might not be able to effectively test a positive diagnostic, but it is good to ensure no diagnostic for the common case.
```suggestion
// Devices that do not support the generic address space shall not specify
// a raw pointer or reference type as the return type or as a parameter type.
[[clang::sycl_external]] int *func0();
[[clang::sycl_external]] int &func1();
[[clang::sycl_external]] int &&func2();
[[clang::sycl_external]] void func3(int *);
[[clang::sycl_external]] void func4(int &);
[[clang::sycl_external]] void func5(int &&);
template<typename T>
[[clang::sycl_external]] void func6(T) {}
template void func6(int *);
template<> [[clang::sycl_external]] void func6<int*>(int *) {}
```

Reply:
`template<> [[clang::sycl_external]] void func6<long*>(long *) {}`